File: src/core/engine.test.ts
import { describe, expect, it } from "vitest";
import emptyTitle from "../../testdata/adversarial/empty-title.json";
import emptyAnswerNonPolicy from "../../testdata/adversarial/empty-answer-non-policy.json";
import looseDate from "../../testdata/adversarial/loose-date.json";
import missingAnswer from "../../testdata/adversarial/missing-answer.json";
import pageZero from "../../testdata/adversarial/page-zero.json";
import whitespaceTitle from "../../testdata/adversarial/whitespace-title.json";
import { SAMPLE_INPUT } from "../fixtures/sample";
import { runPreflight } from "./engine";
import { PreflightError, type PreflightInput } from "./types";

function cloneSample(): PreflightInput {
  return structuredClone(SAMPLE_INPUT);
}

describe("runPreflight", () => {
  it("returns all five checks with deterministic findings", () => {
    const first = runPreflight(cloneSample());
    const second = runPreflight(cloneSample());

    expect(first).toEqual(second);
    expect(first.executedChecks).toEqual([
      "source_coverage",
      "unsupported_claims",
      "conflicts",
      "staleness",
      "must_refuse",
    ]);
    expect(first.findings.some((finding) => finding.type === "must_refuse")).toBe(true);
    expect(
      first.findings.some(
        (finding) => finding.type === "unsupported_claims" && finding.status === "fail",
      ),
    ).toBe(true);
    expect(first.findings.filter((finding) => finding.type === "conflicts" && finding.status === "fail"))
      .toHaveLength(2);
    expect(first.findings.filter((finding) => finding.type === "conflicts").map((finding) => finding.id))
      .toEqual(["conflict:benchmark", "conflict:management_fee"]);
  });

  it("keeps source evidence on support and conflict findings unless inconclusive", () => {
    const report = runPreflight(cloneSample());
    const evidenceFindings = report.findings.filter((finding) =>
      ["source_coverage", "unsupported_claims", "conflicts"].includes(finding.type),
    );

    evidenceFindings.forEach((finding) => {
      if (finding.status !== "inconclusive") {
        expect(finding.sourceRefs.length).toBeGreaterThan(0);
        finding.sourceRefs.forEach((ref) => {
          expect(ref.sourceId).not.toBe("");
          expect(ref.page).toBeGreaterThan(0);
          expect(ref.span).not.toBe("");
        });
      }
    });
  });

  it("fails closed for mixed fundIds", () => {
    const mixed = cloneSample();
    mixed.sources[1].fundId = "BR-OTHER-FUND";

    expect(() => runPreflight(mixed)).toThrowError(PreflightError);
    try {
      runPreflight(mixed);
    } catch (error) {
      expect(error).toMatchObject({ code: "MIXED_FUND_IDS" });
    }
  });

  it("returns inconclusive support when a source span is invalid", () => {
    const invalidRef = cloneSample();
    invalidRef.qas[0].sourceRefs![0].span = "Not present";
    const report = runPreflight(invalidRef);

    expect(report.findings).toContainEqual(
      expect.objectContaining({
        id: "support:qa-management-fee",
        status: "inconclusive",
        sourceRefs: [],
      }),
    );
  });

  it("checks support against exact referenced spans, not unrelated page text", () => {
    const laundered = cloneSample();
    laundered.qas[0].answer = "IMA-B";
    const report = runPreflight(laundered);

    expect(report.findings).toContainEqual(
      expect.objectContaining({
        id: "support:qa-management-fee",
        status: "fail",
      }),
    );
  });

  it.each([
    ["missing answer", missingAnswer],
    ["page zero", pageZero],
    ["empty source title", emptyTitle],
    ["whitespace source title", whitespaceTitle],
    ["empty answer without policy", emptyAnswerNonPolicy],
    ["loose date", looseDate],
  ])("rejects adversarial fixture: %s", (_name, fixture) => {
    expect(() => runPreflight(fixture)).toThrowError(PreflightError);
    try {
      runPreflight(fixture);
    } catch (error) {
      expect(error).toMatchObject({ code: "INVALID_INPUT" });
    }
  });

  it("rejects corrupt and oversized shapes with specific codes", () => {
    expect(() => runPreflight({ schemaVersion: "1.0" })).toThrowError(PreflightError);

    const tooManyQas = cloneSample();
    tooManyQas.qas = Array.from({ length: 21 }, (_, index) => ({
      id: `qa-${index}`,
      question: "Question",
      answer: "Answer",
    }));
    try {
      runPreflight(tooManyQas);
    } catch (error) {
      expect(error).toMatchObject({ code: "LIMIT_EXCEEDED" });
    }

    const tooLarge = cloneSample();
    tooLarge.sources[0].title = "x".repeat(20 * 1024 * 1024);
    try {
      runPreflight(tooLarge);
      throw new Error("Expected oversized input to fail");
    } catch (error) {
      expect(error).toMatchObject({ code: "LIMIT_EXCEEDED" });
    }
  });
});

File: internal/preflight/engine_test.go
package preflight

import (
	"bytes"
	"encoding/json"
	"errors"
	"os"
	"reflect"
	"testing"
)

func loadFixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile("../../testdata/" + name)
	if err != nil {
		t.Fatal(err)
	}
	return data
}

func TestRunReaderDeterministicAllChecks(t *testing.T) {
	data := loadFixture(t, "sample.json")
	first, err := RunReader(bytes.NewReader(data))
	if err != nil {
		t.Fatal(err)
	}
	second, err := RunReader(bytes.NewReader(data))
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(first, second) {
		t.Fatal("same input produced different reports")
	}
	if !reflect.DeepEqual(first.ExecutedChecks, CheckTypes) {
		t.Fatalf("executed checks = %v", first.ExecutedChecks)
	}

	seen := map[string]bool{}
	conflicts := 0
	for _, finding := range first.Findings {
		seen[finding.Type] = true
		if finding.Type == CheckConflicts && finding.Status == "fail" {
			conflicts++
		}
		if (finding.Type == CheckSourceCoverage || finding.Type == CheckUnsupported || finding.Type == CheckConflicts) && finding.Status != "inconclusive" && len(finding.SourceRefs) == 0 {
			t.Fatalf("finding %s has no evidence address", finding.ID)
		}
	}
	if len(seen) != len(CheckTypes) {
		t.Fatalf("got check types %v", seen)
	}
	if conflicts != 2 {
		t.Fatalf("got %d conflicts, want 2", conflicts)
	}
	conflictIDs := make([]string, 0, 2)
	for _, finding := range first.Findings {
		if finding.Type == CheckConflicts {
			conflictIDs = append(conflictIDs, finding.ID)
		}
	}
	if !reflect.DeepEqual(conflictIDs, []string{"conflict:benchmark", "conflict:management_fee"}) {
		t.Fatalf("conflict order = %v", conflictIDs)
	}
}

func TestDecodeInvalidJSON(t *testing.T) {
	_, err := RunReader(bytes.NewReader(loadFixture(t, "invalid.json")))
	var validationError *ValidationError
	if !errors.As(err, &validationError) || validationError.Code != "INVALID_INPUT" {
		t.Fatalf("got %v", err)
	}
}

func TestDecodeOversizedInput(t *testing.T) {
	data := bytes.Repeat([]byte("x"), int(MaxInputBytes+1))
	_, err := RunReader(bytes.NewReader(data))
	var validationError *ValidationError
	if !errors.As(err, &validationError) || validationError.Code != "LIMIT_EXCEEDED" {
		t.Fatalf("got %v", err)
	}
}

func TestMixedFundIDsFailClosed(t *testing.T) {
	_, err := RunReader(bytes.NewReader(loadFixture(t, "mixed-funds.json")))
	var validationError *ValidationError
	if !errors.As(err, &validationError) || validationError.Code != "MIXED_FUND_IDS" {
		t.Fatalf("got %v", err)
	}
}

func TestAdversarialFixturesFailClosed(t *testing.T) {
	fixtures := []string{
		"adversarial/missing-answer.json",
		"adversarial/page-zero.json",
		"adversarial/empty-title.json",
		"adversarial/whitespace-title.json",
		"adversarial/empty-answer-non-policy.json",
		"adversarial/loose-date.json",
	}
	for _, fixture := range fixtures {
		t.Run(fixture, func(t *testing.T) {
			_, err := RunReader(bytes.NewReader(loadFixture(t, fixture)))
			var validationError *ValidationError
			if !errors.As(err, &validationError) || validationError.Code != "INVALID_INPUT" {
				t.Fatalf("got %v", err)
			}
		})
	}
}

func TestSupportUsesExactReferencedSpan(t *testing.T) {
	input, err := Decode(bytes.NewReader(loadFixture(t, "sample.json")))
	if err != nil {
		t.Fatal(err)
	}
	input.QAs[0].Answer = "IMA-B"
	report, err := Run(input)
	if err != nil {
		t.Fatal(err)
	}
	for _, finding := range report.Findings {
		if finding.ID == "support:qa-management-fee" {
			if finding.Status != "fail" {
				t.Fatalf("support status = %s, want fail", finding.Status)
			}
			return
		}
	}
	t.Fatal("support finding not found")
}

func TestOutputIsJSONCompatible(t *testing.T) {
	report, err := RunReader(bytes.NewReader(loadFixture(t, "sample.json")))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := json.Marshal(report); err != nil {
		t.Fatal(err)
	}
}
